### DISCUSSION TITLE
Return a byte array from C++ to react-native in Android

### DISCUSSION THREAD

**Discussion by mrAlmond** [callstack/react-native-builder-bob#454] [Q&A]
Hi everyone,

I'm using this scaffolding project to create a react-native library that must call a low level c++ library to encode/decode messages for a custom serial protocol.
Most of the native C++ functions simply return an "unsigned char *" buffer containing the encoded message bytes.
In iOS I'm able to return this buffer by converting it to a NSData object and then converting it to a NSArray.

In Android, at JAVA level, the return type is mapped by Codegen to WritableArray (the RN spec return value is "number[]").
Then in JNI (cpp-adapter.cpp), where the C++ calls are made, the return value type is "jobject" (I've also tried by using jbyteArray but nothing changes).

**cpp-adapter.cpp**

```
extern "C"
JNIEXPORT jobject JNICALL
Java_com_libcustom_LibcustomModule_nativeLibCustomMakePacket(JNIEnv *env, jclass type) {
    unsigned char *packetBuffer = (unsigned char *)calloc(MAX_BUF_PACK_SIZE, sizeof(unsigned char));
    if( packetBuffer == NULL ) {
        return NULL;
    }

    // Call the native c++ method with some test parameters
    int packetSize = libcustom::libCustomMakePacket(&packetBuffer, 1, 1, 1);
    if( packetSize <= 0 ) {
        free(packetBuffer); // Free allocated buffer
        return NULL;
    }

    // Convert the buffer to a jbyteArray
    auto ret = env->NewByteArray(packetSize);
    if( ret == NULL ) {
        free(packetBuffer); // Free allocated buffer
        return NULL;
    }

    env->SetByteArrayRegion(ret, 0, packetSize, (jbyte *)packetBuffer);
    free(packetBuffer);

    // when ret is returned the application crashes...
    return ret;
}
``` 

**LibcustomModule.java**
```
private static native WritableArray nativeLibCustomMakePacket();

public WritableArray libCustomMakePacket() {
    return nativeLibCustomMakePacket();
}
```

When the call is made the Android app crashes with a SIGABRT.
Other methods returning "number" (mapped to jdouble) works well.
  
In iOS everything works without problems.

Which is the right method to return a byte array from c++ to typescript in Android?

Thank you.

**Comment by satya164** [comment]
Don't know about the method but you'll probably want to convert it into UInt8Array byte array, ArrayBuffer for binary data.

**Reply by mrAlmond** [reply]
If you look at the documentation they said that for native modules there are a limited number of types that are supported.
See here: https://reactnative.dev/docs/native-modules-android#argument-types
Uint8Array is not a supported one...

If I use Uint8Array the Codegen fails with this error:
``
Module NativeLibcustom: Unrecognized generic type 'Uint8Array' in NativeModule spec
``

So I've tried to use number[] or Array\<number\> and in both cases the Codegen maps them to WritableArray (that in the supported types page is not even mentioned, there is only ReadableArray).
This is really strange and confusing...
